examples/cosmos/ics20/ics20.qnt

```diff
@@ -146,90 +219,266 @@ module ics20Test {
+      receivedButUnacknowledgedPackets: Set(),
+      inAcknowledgements: Set()
     })
   }
 
-  /// Receives a packet sent from `sourceChain` to `destChain`
-  action receivePacket(sourceChain: str, destChain: str): bool = {
+  /// Send a packet from `sourceChain` to `destChain`, indicating a transfer
+  /// of `amount`-many tokens of `denom` from `sender` to `receiver`.
+  action sendPacket(sourceChain: str, destChain: str,
+                    denom: DenomTrace, amount: UINT256,
+                    sender: Address, receiver: Address): bool = all {
+    // (1) Pre-condition:
+    // ICS 4: "a channel is a pipeline [...] between specific modules on **separate** blockchains"
+    sourceChain != destChain,
+    // there is a channel between `sourceChain` and `destChain`
+    CHANNEL_TOPOLOGY.keys().contains(sourceChain),
+    CHANNEL_TOPOLOGY.get(sourceChain).keys().contains(destChain),
+
+    // (2) Send the packet using `sendFungibleTokens`:
+    val sourceChainState = chainStates.get(sourceChain)
+    val newSourceChainState = sendFungibleTokens(sourceChainState, denom, amount,
+                                                 sender, receiver,
+                                                 "transfer", CHANNEL_TOPOLOGY.get(sourceChain).get(destChain),
+                                                 0, 0)
+
+    // (3) Quint state transition:
+    chainStates' = chainStates.set(sourceChain, newSourceChainState)
+  }
+
+  /// Receive a packet sent from `sourceChain` to `destChain`:
+  /// - Pick an arbitrary, previously unreceived packet sent from `sourceChain`
+  ///   to `destChain`.
+  /// - Call the `onRecvPacket` callback on it.
+  /// - Update the chain states to record the packet as received.
+  /// - Record (but not receive) the acknowledgement produced by `onRecvPacket`
+  ///   on `destChain`.
+  action receivePacket(sourceChain: str, destChain: str): bool =
     val sourceChainState = chainStates.get(sourceChain)
-    nondet packet = sourceChainState.outPackets.oneOf()
-    val result = onRecvPacket(chainStates.get(destChain), packet)
     all {
-      chainStates' = chainStates
-        .set(destChain, result._1)
-        .set(sourceChain, sourceChainState.with("outPackets", sourceChainState.outPackets.exclude(Set(packet))))
+      // (1) Pre-condition:
+      // ICS 4: "a channel is a pipeline [...] between specific modules on **separate** blockchains"
+      sourceChain != destChain,
+      // there is a channel between `sourceChain` and `destChain`
+      CHANNEL_TOPOLOGY.keys().contains(sourceChain),
+      CHANNEL_TOPOLOGY.get(sourceChain).keys().contains(destChain),
+      // there is an unreceived packet traveling from `sourceChain` to `destChain`
+      sourceChainState.outPackets.exists(packet => packet.isTravelingBetween(sourceChain, destChain)),
+
+      // (2) Non-deterministically pick a packet from `sourceChain` to receive on `destChain`:
+      nondet packet = sourceChainState.outPackets.oneOf()
```

Comment:
Just a note, you might want to change this or not: It is more efficient for the random simulator if you filter the set before calling `oneOf` (as oppose to having the `isTravelingBetween` condition being enforced on the picked value). What will happen here is that, if there are too many packets in `outPackets`, the simulator will try to pick values a few times and then fail the action if the randomly picked values are not the correct ones to satisfy the condition. Filtering makes it that the the simulator always gets this right.

This problem might not be relevant for the simulations we are currently doing, but I had issues with this in the `denomTrace` spec and decided do make sure I'm always filtering my sets before calling `oneOf` on them.

Reply:
Right. This may show some difference in the simulator, as it would reject some traces, if the condition is checked after `oneOf`. If you filter a set upfront, before calling `oneOf`, then the simulator would not be able to make an infeasible choice. This would not matter for Apalache though.

Reply:
Aaah yes, this also looks nicer! 0dba89c
I actually tried to write a predicated `oneOf` before writing this. Minimality of the core language aside, would it make sense to have `oneOf` optionally take a filter?